packages/tool-cache/src/tool-cache.ts

```diff
@@ -81,24 +97,43 @@ export async function downloadTool(
         throw err
       }
 
+      let error: Error | undefined
       const file: NodeJS.WritableStream = fs.createWriteStream(dest)
       file.on('open', async () => {
-        try {
-          const stream = response.message.pipe(file)
-          stream.on('close', () => {
-            core.debug('download complete')
-            resolve(dest)
-          })
-        } catch (err) {
-          core.debug(
-            `Failed to download from "${url}". Code(${response.message.statusCode}) Message(${response.message.statusMessage})`
-          )
-          reject(err)
-        }
+        const responseMessageFactory = getGlobal<() => stream.Readable>(
+          'TEST_DOWNLOAD_TOOL_RESPONSE_MESSAGE_FACTORY',
+          () => response.message
+        )
+        const readStream = responseMessageFactory()
+        readStream.pipe(file)
```

Comment:
Thoughts on using [`pipeline`](https://nodejs.org/api/stream.html#stream_stream_pipeline_source_transforms_destination_callback)? It can be promisified, and deals with errors from all streams

Reply:
will look into switching over to that

Reply:
that is waaaaay better

thanks!

Reply:
great! 👍